diff --git a/CMakeLists.txt b/CMakeLists.txt
index fcb8103..c1f3967 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -18,7 +18,11 @@ set(CXX_FLAGS
  -Wextra
  -Werror
  -Wconversion
+ -Wno-null-dereference
+ -Wno-sign-conversion
+ -Wno-unused-local-typedef
  -Wno-unused-parameter
+ -Wno-unused-private-field
  -Wold-style-cast
  -Woverloaded-virtual
  -Wpointer-arith
@@ -27,16 +31,15 @@ set(CXX_FLAGS
  -march=native
  # -MMD
  # -std=c++0x
- -rdynamic
  )
 if(CMAKE_BUILD_BITS EQUAL 32)
   list(APPEND CXX_FLAGS "-m32")
 endif()
 string(REPLACE ";" " " CMAKE_CXX_FLAGS "${CXX_FLAGS}")
 
-set(CMAKE_CXX_COMPILER "g++")
+set(CMAKE_CXX_COMPILER "clang++")
 set(CMAKE_CXX_FLAGS_DEBUG "-O0")
-set(CMAKE_CXX_FLAGS_RELEASE "-O2 -finline-limit=1000 -DNDEBUG")
+set(CMAKE_CXX_FLAGS_RELEASE "-O2 -DNDEBUG")
 set(EXECUTABLE_OUTPUT_PATH ${PROJECT_BINARY_DIR}/bin)
 set(LIBRARY_OUTPUT_PATH ${PROJECT_BINARY_DIR}/lib)
 
