diff --git a/CMakeLists.txt b/CMakeLists.txt
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -15,6 +15,7 @@
  -Wextra
  -Werror
  -Wconversion
+ -Wno-sign-conversion
  -Wno-unused-parameter
  -Wold-style-cast
  -Woverloaded-virtual
@@ -24,16 +25,15 @@
  -march=native
  # -MMD
  # -std=c++0x
- -rdynamic
  )
 if(CMAKE_BUILD_BITS EQUAL 32)
   list(APPEND CXX_FLAGS "-m32")
 endif()
 string(REPLACE ";" " " CMAKE_CXX_FLAGS "${CXX_FLAGS}")
 
-set(CMAKE_CXX_COMPILER "g++")
+set(CMAKE_CXX_COMPILER "clang++")
 set(CMAKE_CXX_FLAGS_DEBUG "-O0")
-set(CMAKE_CXX_FLAGS_RELEASE "-O2 -finline-limit=1000 -DNDEBUG")
+set(CMAKE_CXX_FLAGS_RELEASE "-O2 -DNDEBUG")
 set(EXECUTABLE_OUTPUT_PATH ${PROJECT_BINARY_DIR}/bin)
 set(LIBRARY_OUTPUT_PATH ${PROJECT_BINARY_DIR}/lib)
 
@@ -55,6 +55,8 @@
 endif()
 
 include_directories(${Boost_INCLUDE_DIRS})
+include_directories(/usr/lib/gcc/x86_64-linux-gnu/4.4/include/)
+include_directories(/usr/lib/gcc/x86_64-linux-gnu/4.4/include-fixed/)
 
 include_directories(${PROJECT_SOURCE_DIR})
 
